toolkit/tools/downloader/downloader.go

```diff
@@ -0,0 +1,130 @@
+	tlsClientKey  = app.Flag("private-key", "TLS client key to use when downloading files.").String()
+
+	dstFile   = app.Flag("output-file", "Destination file to download to").Short('O').String()
+	prefixDir = app.Flag("directory-prefix", "Directory to download to").Short('P').String()
+	srcUrl    = app.Arg("url", "URL to download").Required().String()
+)
+
+func main() {
+	app.Version(exe.ToolkitVersion)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+	logger.InitBestEffort(*logFile, *logLevel)
+	if *noVerbose {
+		logger.Log.SetLevel(logrus.ErrorLevel)
+	}
+
+	if *noClobber {
+		exists, err := file.PathExists(*dstFile)
+		if err != nil {
+			logger.Log.Fatalf("Failed to check if file (%s) exists. Error: %s", *dstFile, err)
+		}
+		if exists {
+			logger.Log.Infof("File (%s) already exists, skipping download.", *dstFile)
+			return
+		}
+	}
+
+	caCerts, err := x509.SystemCertPool()
+	if err != nil {
+		logger.Log.Fatalf("Failed to load system certificate pool. Error: %s", err)
+	}
+	if *caCertFile != "" {
+		newCACert, err := os.ReadFile(*caCertFile)
+		if err != nil {
+			logger.Log.Fatalf("Invalid CA certificate (%s), error: %s", *caCertFile, err)
+		}
+
+		caCerts.AppendCertsFromPEM(newCACert)
+	}
+
+	var tlsCerts []tls.Certificate
+	if *tlsClientCert != "" && *tlsClientKey != "" {
+		cert, err := tls.LoadX509KeyPair(*tlsClientCert, *tlsClientKey)
+		if err != nil {
+			logger.Log.Fatalf("Invalid TLS client key pair (%s) (%s), error: %s", *tlsClientCert, *tlsClientKey, err)
+		}
+
+		tlsCerts = append(tlsCerts, cert)
+	}
+
+	// dst may be empty, in which case the file will be downloaded to the current directory. Generate dst from src's basename.
+	// The url may include query strings which should be stripped.
+	if *dstFile == "" {
+		// strip query strings from url
+		u, err := url.Parse(*srcUrl)
+		if err != nil {
+			logger.Log.Fatalf("Invalid URL (%s), error: %s", *srcUrl, err)
+		}
+		*dstFile = filepath.Base(u.Path)
+		if *prefixDir != "" {
+			*dstFile = filepath.Join(*prefixDir, *dstFile)
+		}
+	} else if *prefixDir != "" {
+		logger.Log.Fatalf("Cannot specify both --output-file and --directory-prefix")
+	}
+
+	err = downloadFile(*srcUrl, *dstFile, caCerts, tlsCerts)
+	if err != nil {
+		logger.Log.Fatalf("Failed to download (%s) to (%s). Error: %s", *srcUrl, *dstFile, err)
+	}
+}
+
+func downloadFile(srcUrl, dstFile string, caCerts *x509.CertPool, tlsCerts []tls.Certificate) (err error) {
+	const (
+		// With 6 attempts, initial delay of 1 second, and a backoff factor of 3.0 the total time spent retrying will be
+		// 1 + 3 + 9 + 27 + 81 = 121 seconds.
+		downloadRetryAttempts = 6
+		failureBackoffBase    = 3.0
+		downloadRetryDuration = time.Second
+	)
+	var noCancel chan struct{} = nil
+
+	_, err = retry.RunWithExpBackoff(func() error {
+		netErr := network.DownloadFile(srcUrl, dstFile, caCerts, tlsCerts)
+		if netErr != nil {
+			logger.Log.Warnf("Failed to download (%s). Error: %s", srcUrl, netErr)
```

Comment:
nit: having same style of error logging throughout the file would be nice.
'. Error:'   v/s    ', error:'     v/s     '. Error:\n'

Reply:
I'd like to keep these intentionally as one line, (Also, the logger can't use %w).
I can add newlines to the fatals though.